worker/instancepoller/worker.go

```diff
@@ -85,57 +180,319 @@ func (u *updaterWorker) Wait() error {
+	}
+
+	// Compare the addreses reported by the provider with the ones recorded
+	// at the machine model and trigger an update if required.
+	curAddresses, err := entry.m.ProviderAddresses()
+	if err != nil {
+		return status.Unknown, errors.Trace(err)
+	}
+
+	providerAddresses, err := info.Addresses(u.callContext)
+	if err != nil {
+		return status.Unknown, errors.Trace(err)
+	}
+
+	if !addressesEqual(curAddresses, providerAddresses) {
+		u.config.Logger.Infof("machine %q (instance ID %q) has new addresses: %v", entry.m.Id(), entry.instanceID, providerAddresses)
+		if err := entry.m.SetProviderAddresses(providerAddresses...); err != nil {
+			u.config.Logger.Errorf("cannot set addresses on %q: %v", entry.m, err)
+			return status.Unknown, errors.Trace(err)
+		}
+	}
+
+	return providerStatus.Status, nil
 }
 
-// dying is part of the lifetimeContext interface.
-func (u *updaterWorker) dying() <-chan struct{} {
-	return u.catacomb.Dying()
+func (u *updaterWorker) maybeSwitchPollGroup(curGroup pollGroupType, entry *pollGroupEntry, curProviderStatus, curMachineStatus status.Status) {
+	if curProviderStatus == status.Allocating || curProviderStatus == status.Pending {
+		// Keep the machine in the short poll group until it settles
+		entry.bumpShortPollInterval(u.config.Clock)
+		return
+	}
+
+	machAddrs, _ := entry.m.ProviderAddresses()
+
+	// If the machine is currently in the long poll group and it has an
+	// unknown status or suddenly has no network addresses, move it back to
+	// the short poll group.
+	if curGroup == longPollGroup && (curProviderStatus == status.Unknown || len(machAddrs) == 0) {
+		delete(u.pollGroup[longPollGroup], entry.tag)
+		u.pollGroup[shortPollGroup][entry.tag] = entry
+		entry.resetShortPollInterval(u.config.Clock)
+		u.config.Logger.Debugf("moving machine %q (instance ID %q) back to short poll group", entry.m, entry.instanceID)
+		return
+	}
+
+	// The machine has started and we have at least one address; move to
+	// the long poll group
+	if len(machAddrs) > 0 && curMachineStatus == status.Started {
+		if curGroup == longPollGroup {
+			return // already in long poll group
+		}
+		delete(u.pollGroup[shortPollGroup], entry.tag)
+		u.pollGroup[longPollGroup][entry.tag] = entry
+		u.config.Logger.Debugf("moving machine %q (instance ID %q) to long poll group", entry.m, entry.instanceID)
+		return
+	}
+
+	// If we are in the short poll group apply exponential backoff to the
+	// poll frequency allow time for the machine to boot up.
+	if curGroup == shortPollGroup {
+		entry.bumpShortPollInterval(u.config.Clock)
+	}
 }
 
-// errDying is part of the lifetimeContext interface.
-func (u *updaterWorker) errDying() error {
-	return u.catacomb.ErrDying()
+// addressesEqual compares the addresses of the machine and the instance information.
+func addressesEqual(addrs0, addrs1 network.ProviderAddresses) bool {
+	if len(addrs0) != len(addrs1) {
+		return false
+	}
+
+nextAddr:
+	for _, a0 := range addrs0 {
```

Comment:
in weird scenarios this isn't quite right, but it would require duplicated addresses (eg, 1, 2, 2, 3 would match 1, 2, 3, 4, because the lengths are the same, and every entry in addrs0 is found in addrs1).
As Joe mentioned, I think we have code like this, and it would be better to use shared code, because it has taken a few tries to get right with things like space names, ordering, etc getting into the mix.

Reply:
This code was copied over from the original implementation. I will remove this and try to find the newer comparison code that you and Joe mentioned.